### PR TITLE
Small sampling optimizations

### PR DESCRIPTION
**Pull request by kalomaze**
I pointed out in https://github.com/ggerganov/llama.cpp/discussions/5073 that there seems to be some not completely insignificant overhead with current sampling operations:

image

With the change I made in this PR, I managed to reduce 15ms per token (for sampling over the whole initial vocabulary) to 11ms per token:

`llama_print_timings:      sample time =   11927.89 ms /  1024 runs   (   11.65 ms per token,    85.85 tokens per second)`

Efficient top k sorting is probably next.

### CODE REVIEW

**Comment by kalomaze** [comment]
image
...are my eyes deceiving me or did the prompt evaluation speed positively improve as a result of this? Shouldn't that phase be disconnected from the sample softmax function entirely? Or perhaps it's some weird CPU scheduling thing?

File: llama.cpp [hunk 1]
```diff
@@ -7973,23 +7973,26 @@ void llama_sample_softmax(struct llama_context * ctx, llama_token_data_array * c
...
-    const int64_t t_start_sample_us = ggml_time_us();
+    if (candidates->sorted) {
+       candidates->size = k;
+       return;
+    }
 
+    const int64_t t_start_sample_us = ggml_time_us();
+    
     k = std::max(k, (int) min_keep);
     k = std::min(k, (int) candidates->size);
```
Comment (kalomaze) [inline llama.cpp:7984]: I'm realizing now that these sanity checks will not run if we return early. Does that matter?

**Comment by ggerganov** [comment]
Which change exactly results in the improvement? Most of these are noops, except the temperature check - but I don't think it can amount to that amount

**Comment by kalomaze** [comment]
> Which change exactly results in the improvement? Most of these are noops, except the temperature check - but I don't think it can amount to that amount

Calculating the exp and sum in one pass, in the softmax function

**Comment by ggerganov** [comment]
I don't get it - it's the same thing before and after, except a temporary variable. It cannot have any effect on the performance

**Comment by kalomaze** [comment]
> I don't get it - it's the same thing before and after, except a temporary variable. It cannot have any effect on the performance

I might have tricked myself into thinking it did something because of a build flag or something. Prompt eval speed is consistent if I abandon the commits and rebuild from scratch, so it's probably that

File: llama.cpp [hunk 2]
```diff
@@ -8027,14 +8033,14 @@ void llama_sample_top_p(struct llama_context * ctx, llama_token_data_array * can
...
 
 void llama_sample_min_p(struct llama_context * ctx, llama_token_data_array * candidates, float p, size_t min_keep) {
-    if (p <= 0.0f || !candidates->size) {
+    if (p <= 0.0f) {
         return;
     }
+    
+    const int64_t t_start_sample_us = ggml_time_us();
 
     llama_sample_softmax(ctx, candidates);
```
Comment (ggerganov) [inline llama.cpp:8043]: This will count the `llama_sample_softmax()` time 2 times. Either change to `llama_sample_softmax(nullptr, candidates);` or keep it before the `ggml_time_us()` call

File: llama.cpp [hunk 3]
```diff
@@ -8118,11 +8125,11 @@ void llama_sample_typical(struct llama_context * ctx, llama_token_data_array * c
         return;
     }
 
+    const int64_t t_start_sample_us = ggml_time_us();
+
     // Compute the softmax of logits and calculate entropy
     llama_sample_softmax(nullptr, candidates);
```
Comment (ggerganov) [inline llama.cpp:8131]: This is OK because it does not take a context

File: llama.cpp (hunk 1 above)
Comment (ggerganov) [reply llama.cpp:7984]: I think they need to remain - simply move them before the "if sorted" early return at the top

**Comment by kalomaze** [comment]
Ok, this is odd.

If topk=31999:
`llama_print_timings:      sample time =    5220.53 ms /   512 runs   (   10.20 ms per token,    98.07 tokens per second)`
If topk=32000:
`llama_print_timings:      sample time =    8418.14 ms /   512 runs   (   16.44 ms per token,    60.82 tokens per second)`

1.6x faster if we sort _almost_ the whole vocabulary, but not exactly?

**Review by ggerganov — APPROVED** [review]

**Comment by kalomaze** [comment]
image
With Repetition Penalty on and topk=32000 (so, the worst case scenario), ~30% of my generation time is spent on sampling for 7b q8_0.
I'm thinking that the ideal solution may be to move the sampling operations to GPU to minimize idle time

**Comment by ggerganov** [comment]
But what's the point of `topk = 32000` - it's a noop since all candidates remain

**Comment by kalomaze** [comment]
> But what's the point of `topk = 32000` - it's a noop since all candidates remain

topk=32000 is equivalent in terms of speed to topk=0 (which defaults to the full vocabulary in the same way), so something is presumably up with how the softmax function sorts / does sort checks later on?

image

**Comment by cmp-nct** [comment]
I didn't notice this was already in work.
I found kalomaze's post strange as in 100 times slowness, it's the top-k sorting not the softmax.

80-90% of the large-k latency comes from sorting/partial sorting. Anything that comes after is almost irrelevant.
Check my PR, it's about 30% faster - maybe more for a slower CPU
https://github.com/ggerganov/llama.cpp/pull/5085

The real way forward imho is a dyn-k function that takes in the sampling settings and estimates a proper top-k, then the partial sort will cause a 2-3 times reduction in latency.

**Comment by kalomaze** [comment]
I believe this PR is made redundant by the other work done on topk/minp sorting optimizations.
